### DISCUSSION TITLE
About disabled buttons variants

### DISCUSSION THREAD

**Discussion by parafeu** [saadeghi/daisyui#516] [Ideas / feature requests]
Hey ! 

Starting this discuss after my recent issue #515  (which was not one finally).

I find it strange to lose the variation of a button when it is disabled. 
In my opinion, the color variation on a button are mainly used to indicate a difference in use. 

Here is an exemple : 

![image](https://user-images.githubusercontent.com/39409783/155539470-2b533193-2f8e-4cc2-af67-cb908dc9cba9.png)

On this UI, the button with the primary variation represents the primary choice while the button with the ghost variation is the secondary choice.

![image](https://user-images.githubusercontent.com/39409783/155539810-42ed91c8-273a-493c-952e-4bb611efcd5d.png)

But when the primary button is disabled (because the form is not filled), the information that it is still the primary option is lost.

![image](https://user-images.githubusercontent.com/39409783/155540548-a6a710c8-ab42-4701-a162-be357e880bc5.png)

This is what I think a disabled button should look like. We understand that it is impossible to use, but the information provided by the variation is not lost. 

It's my opinion of course, that's why I'm opening that discuss.

And I hope you'll understand what is say, my english is very far from perfect 😅.

**Comment by saadeghi** [comment]
I don't think a disabled `primary` button should look different from a `secondary` button. This is not a fact but my reason is this:  

> The reason we use buttons with different colors in the first place is to make it easier for users to distinguish actions from each other when they want to click them. 
If we put 3 gray buttons next to each other, users need to actually read all the buttons, then decide which one to click then click. 
For a better UX, we use 2 gray buttons and 1 green button to encourage users to click on the most important button without having to decide.
Read this: https://www.oreilly.com/library/view/ux-for-beginners/9781491912676/ch67.html

When it comes to disabled buttons, there's no choice anymore. We just want to tell the user that there's a button there but it can't be clicked.

Anyway... If you think you need a different color for disabled button, It's just 1 line of CSS to customize it:
```css
.btn{
  @apply disabled:bg-red-300 disabled:text-blue-500;
}
```
example: https://play.tailwindcss.com/zIrat7gSC1?file=css

**Comment by parafeu** [comment]
Personally I have never seen a design system where the disabled buttons lose their colors. 

Whether it's on bootstrap or Material for example, the disabled buttons keep their variation.

I'm not saying that daisyui should absolutely follow these examples of course, but I think that making a disabled button lose its variation is an unnecessary limitation. 


Here is the code needed to modify all the button variations so that they do not lose their color :
```css
.btn-primary {
  @apply disabled:bg-primary;
}

.btn-secondary {
  @apply disabled:bg-secondary;
}

.btn-accent {
  @apply disabled:bg-accent;
}

.btn-info {
  @apply disabled:bg-info;
}

.btn-success {
  @apply disabled:bg-success;
}

.btn-warning {
  @apply disabled:bg-warning;
}

.btn-error {
  @apply disabled:bg-error;
}
```

While if the buttons kept the color, the code to switch them all to the same color if disabled is shorter.

```css
.btn {
  @apply disabled:!bg-neutral disabled:!text-neutral-content;
}
```

In my opinion, this makes the design system more flexible.

**Comment by saadeghi** [comment]
> Personally I have never seen a design system where the disabled buttons lose their colors.

I created codesandbox demos for 4 most popular design systems: 
- Material UI: https://codesandbox.io/s/containedbuttons-material-demo-forked-duivmn?file=/demo.js
- Carbon Design System: https://codesandbox.io/s/hopeful-morning-yj16eg?file=/index.js
- Atlassian Design system: https://codesandbox.io/s/button-disabled-forked-cuxs5e?file=/example.tsx
- Ant Design: https://codesandbox.io/s/disabled-antd-4-18-8-forked-lmldm9?file=/index.js

Feel free to check for more: https://github.com/saadeghi/design-systems

**Comment by carlosvigil** [comment]
The conclusion I can come to here is that in the undesirable example, the styling of the button leads to a loss of contrast. Contrast that is kept in the examples @saadeghi provided. I.e. make sure those two styles don't lead to this result, (bg & disabled bg being the same)
